q2_metadata/normalization/_norm_check_rules.py

```diff
@@ -0,0 +1,311 @@
+        # checks that all the items in the expected value list are strings.
+        str_items = [x for x in rule_value if isinstance(x, str)]
+        if len(str_items) == len(rule_value):
+            return True
+    return False
+
+
+def check_ontology(rule_value):
+    """Check that the user-defined value for the
+    rule "ontology" is correctly formatted.
+
+    ontology must be a str
+        Name of the ontology to lookup within for control vocabulary, e.g.
+                    ontology: Gazetteer
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "ontology".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    if isinstance(rule_value, str):
+        if rule_value in ['Gazetteer ontology']:
+            return True
+    return False
+
+
+def check_remap(rule_value):
+    """Check that the user-defined value for the
+    rule "remap" is correctly formatted.
+
+    remap must be a dict
+        Apply the rule of remapping variables values: replace
+        all instances of a given value by another one value.
+                    remap:
+                      str: str
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "remap".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    if isinstance(rule_value, dict):
+        str_values = [
+            x for x, y in rule_value.items() if
+                isinstance(x, (str, int, float)) and
+                isinstance(y, (str, int, float))
+        ]
+        if len(str_values) == len(rule_value.keys()):
+            return True
+    return False
+
+
+def check_validation(rule_value):
+    """Check that the user-defined value for the
+    rule "validation" is correctly formatted.
+
+    validation must be a dict
+        Apply the rule of validating variables values: replace by the blank
+        (or the missing) value every sample entry that for the current variable has
+        in one or multiple other variable(s) a null or missing value for that sample.
+                    validation:
+                      force_to_blank_if:
+                        is null:
+                        - variable
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "validation".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    # accepted structure for the "validation" rule
+    possible_validations = {
+        'force_to_blank_if': {'is null'}
+    }
+    # check the first level is a dict
+    if isinstance(rule_value, dict):
+        # get the current keys
+        rule_value_set = set(rule_value)
+        # get the accepted keys from the current keys
+        accepted_set = rule_value_set & set(possible_validations)
+        # if there is not more keys than the accepted ones
+        if not len(rule_value_set ^ accepted_set):
```

Comment:
Is a set symmetric difference the correct operation? The result of the ^ would be the items in rule_value_set that are not in accepted_set combined with the items in accepted_set that are not in rule_value_set. And then the set cardinality is computed, and we enter the if-statement when that condition is False rather than equal to 0.

Isn't this equivalent to if `rule_value_set.isdisjoint(accepted_set)`? If so, then that may simplify the code, but I'm having a hard time understanding the interpretation of this.

It looks like the subsequent statement uses things from `possible_validations` which must exist in `accepted_set` given line 192, as well as things from `rule_value` which cannot be in the `accepted_set` per the if-statement on like 194. But the indexing performed on line 196 and 198 suggests that it doesn't matter what other keys are in `possible_validations` or `rule_value`, so in fact as the code is written, none of this matters and this function can be replaced with:

`return any([rule_2 in possible_validations['force_to_blank_if'] for rule_2 in rule_value['force_to_blank_if'])`

...with I think the exception of an edge case on `rule_value_2` possibly not being a list?

I'm not convinced anyone knows what this function is actually supposed to do, which is bad.

Reply:
this function was a little complex because it is meant to check the valid formatting of a two-levels nested disctionary ending up with a list. Note that it is ver likely that the allowed keys at the different levels may be updated, e.g. for now we develop for:
```
  force_to_blank_if:
    is null:
    - metadata_variable_x
    - metadata_variable_y
```
but it could well be expanded to:
```
  force_to_blank_if:
    is null:
    - metadata_variable_x
    - metadata_variable_y
    is zero:
    - metadata_variable_x
    - metadata_variable_y
  etc..
```

here the generic function for that format checking:
```
        if isinstance(rule_value, dict):
            if set(rule_value).issubset(allowed):
                for allowed_key, nested_dict in rule_value.items():
                    if isinstance(nested_dict, dict):
                        for nested_rule, variables in nested_dict.items():
                            if nested_rule not in allowed[allowed_key]:
                                return ['not allowed', nested_rule]
                            if not isinstance(variables, list):
                                return ['not a list', variables]
                    else:
                        return ['not a nested dictionary', nested_dict]
            else:
                not_allowed = '", "'.join(set(rule_value).difference(allowed))
                return ['not allowed', not_allowed]
        else:
            return ['not a dictionary', rule_value]
        return []
```